### PR TITLE
cli: Initial Oasis CLI

### PR DESCRIPTION
**Pull request by kostko**

### CODE REVIEW

**Comment by pro-wh** [comment]
going to rebase and merge!

**Review by matevz — COMMENTED** [review]
First pass, some minor questions.

File: cli/cmd/contracts.go
Comment (kostko) [reply cli/cmd/contracts.go:31]: Currently yes, but in theory for any ParaTime that includes the `contracts` module.

File: cli/config/network.go
Comment (kostko) [reply cli/config/network.go:70]: Nice catch, yes we definitely should!

File: client-sdk/go/crypto/signature/ed25519/signer.go
Comment (pro-wh) [inline client-sdk/go/crypto/signature/ed25519/signer.go:30]: could alternatively make the signer field public

could alternatively store the consensus signer and wrap on demand

**Review by pro-wh — COMMENTED** [review]
15/36

File: cli/cmd/wallet.go
Comment (pro-wh) [inline cli/cmd/wallet.go:203]: passphrase should already be a string here

File: cli/cmd/wallet.go
Comment (pro-wh) [inline cli/cmd/wallet.go:175]: here's a part that we could make friendlier imo. this question depends on what what you chose for 'kind,' but we cram it into a universal 'data' question. if we already knew, we could give more detailed instructions in the prompt, e.g. "mnemonic" for mnemonic imports, we could additionally ask for account number, or we could say that the private key must be base64

maybe we could split this off into a separate `Ask`

File: cli/cmd/wallet.go
Comment (pro-wh) [inline cli/cmd/wallet.go:179]: :information_source: the passphrase is not something used to protect the derivation of the account key. this is a passphrase used to seal the account key for file-based wallets

File: cli/wallet/file/ed25519.go
Comment (pro-wh) [inline cli/wallet/file/ed25519.go:11]: what do we use in the relatively raw signers in oasis-core, if not raw ed25519 keys?

File: cli/wallet/file/file.go
Comment (pro-wh) [inline cli/wallet/file/file.go:63]: if anyone wants to bikeshed, tell us your favorite alternative

**Review by pro-wh — COMMENTED** [review]
33/36

File: cli/wallet/file/file.go
Comment (pro-wh) [inline cli/wallet/file/file.go:176]: could we panic on these, if you'd rather not propagate the errors

the implementation of NewEntropy looks like it returns unrandomized bytes if rand.Read fails. we handle rand.Read failures elsewhere in this program.

File: cli/wallet/file/file.go
Comment (pro-wh) [inline cli/wallet/file/file.go:34]: seeing as the Algorithm field covers the derivation algorithm and the resulting key type, maybe this should be ed25519-adr0008?

Comment (pro-wh) [reply cli/wallet/file/file.go:34]: or no, this is used in both walletConfig and secretState, which kind of conflates the meaning.

when walletConfig.Algorithm is algorithmEd25519, it means "loading this wallet will give you an ed25519 key." but it's stored next to a Number field, which is rather something that controls a derivation not hinted at by this field

when secretState.Algorithm is algorithmEd25519, it means "use ADR 0008 to derive an ed25519 key from the mnemonic." it's stored next to that mnemonic.

some weirdness from all this:

- the choice between ADR 0008 derivation and raw private key is within the sealed data, but the account number for the ADR 0008 case is out in the open
- there's this piece of code that switches around algorithmEd25519Raw based on a field in ImportSource

File: cli/wallet/file/ed25519.go
Comment (pro-wh) [reply cli/wallet/file/ed25519.go:11]: and if not 'raw ed25519 keys,'

1. why make the CLI use raw ed25519 keys?
2. shall we instead add this functionality directly to oasis-core?

File: cli/wallet/file/file.go
Comment (pro-wh) [inline cli/wallet/file/file.go:245]: we use a special value here instead of, say storing the ImportSource.kind alongside. what does this get us? maybe in the future we can change the default derivation scheme used when the user only specifies "mnemonic"?

File: cli/wallet/file/file.go
Comment (pro-wh) [inline cli/wallet/file/file.go:43]: I think this is never assigned. how do you envision this being populated in the future? fed in through a new field in ImportSource maybe?

**Review by pro-wh — CHANGES_REQUESTED** [review]
requesting changes: error handling when obtaining entropy

other notes: I want to learn about the higher level architecture of the wallet data model, I've added questions inline

File: cli/wallet/file/ed25519.go
Comment (kostko) [reply cli/wallet/file/ed25519.go:11]: > what do we use in the relatively raw signers in oasis-core (file, memory), if not raw ed25519 keys?

We do, but they cannot be used here as they don't support loading the raw private key value (you can only generate a new random key or generate one from a RFC 8032 seed).

File: cli/wallet/file/file.go
Comment (kostko) [reply cli/wallet/file/file.go:176]: Yeah these should definitely propagate errors, not sure how this got missed.

File: cli/wallet/file/file.go
Comment (Yawning) [reply cli/wallet/file/file.go:63]: Argon2

File: cli/config/config.go
Comment (kostko) [reply cli/config/config.go:56]: Yeah there's a bunch of issue reports upstream (https://github.com/mitchellh/mapstructure/issues/267, https://github.com/mitchellh/mapstructure/issues/166). It is not related to serialization formats as this happens without any serialization (transforming a typed struct into `map[string]interface{}`).

File: cli/wallet/file/file.go
Comment (kostko) [reply cli/wallet/file/file.go]: Yeah that's ugly, I've switched this over so that there's now `ed25519-adr8` and `ed25519-raw` in the open. The number field is indeed something that is specific to ADR8 derivation though.

File: cli/cmd/wallet.go
Comment (kostko) [reply cli/cmd/wallet.go]: The prompt has been changed to be more clear (also added repeat passphrase prompt).

File: cli/wallet/file/ed25519.go
Comment (pro-wh) [reply cli/wallet/file/ed25519.go:11]: would you be opposed to adding that import functionality to the memory signer?

File: cli/config/config.go
Comment (pro-wh) [reply cli/config/config.go:56]: ah thanks

File: cli/cmd/wallet.go
Comment (pro-wh) [reply cli/cmd/wallet.go]: nice, thanks

**Review by pro-wh — APPROVED** [review]

**Comment by pro-wh** [comment]
thanks for the changes!

File: cli/wallet/file/ed25519.go
Comment (kostko) [reply cli/wallet/file/ed25519.go:11]: I wouldn't be, but I would rather avoid the need for this PR to depend on an unreleased version of Oasis Core. Can you raise an issue in Oasis Core (maybe also for exposing the chain context separator) and then once it is part of a stable release we can change it here.

File: cli/cmd/common/transaction.go
Comment (kostko) [reply cli/cmd/common/transaction.go:272]: Good point, the whole `WaitForEvent` should return `interface{}` instead.

File: cli/cmd/common/transaction.go
Comment (kostko) [reply cli/cmd/common/transaction.go:259]: They get closed after the context is cancelled.

Comment (kostko) [reply cli/cmd/common/transaction.go:259]: I'll add a subcontext that will get cancelled as soon as an event is matched.

File: cli/cmd/common/transaction.go
Comment (kostko) [reply cli/cmd/common/transaction.go:270]: Updated the comment to mention this.

File: cli/cmd/common/transaction.go
Comment (kostko) [reply cli/cmd/common/transaction.go]: Yeah, fixed.

File: client-sdk/go/modules/contracts/contracts.go
Comment (kostko) [reply client-sdk/go/modules/contracts/contracts.go]: True, refactored this.

File: cli/main.go
Comment (tjanez) [inline cli/main.go:1]: Could you add an initial `README.md` to the `cli` directory and explain how people can build and start using the Oasis CLI?

File: cli/cmd/root.go
Comment (tjanez) [inline cli/cmd/root.go:1]: Could you add a `version` sub-command or a `--version` top-level flag?

**Comment by Yawning** [comment]
Bikeshedding this to hell and back is great and all, but it has utility functions that I need for other work.

File: cli/cmd/root.go
Comment (kostko) [reply cli/cmd/root.go:1]: Added.

File: cli/main.go
Comment (kostko) [reply cli/main.go:1]: Added something basic.

File: cli/go.mod
Comment (tjanez) [inline cli/go.mod:17]: Could you update cobra to 1.3.0? It has some completion sub-command improvements.

File: cli/main.go
Comment (tjanez) [reply cli/main.go:1]: It looks very good, thanks!

File: cli/cmd/root.go
Comment (tjanez) [reply cli/cmd/root.go:1]: Thanks!

File: .gitignore
Comment (tjanez) [inline .gitignore:27]: Also add `cli/oasis`

File: cli/cmd/network.go
Comment (tjanez) [inline cli/cmd/network.go:1]: Could we add a (sub)command to `network` command or a new command that would show you the current default network and/or the current default ParaTime?

Or perhaps the default network could be shown via a `(default)` suffix (or similar) when running `oasis network list`?

File: cli/cmd/network.go
Comment (tjanez) [inline cli/cmd/network.go:89]: Could we validate user's input before asking survey questions?
I.e. running `net.Validate()` with just the given chain context and RPC so that users aren't asked those questions if we already know that the command will fail.

To prevent things like:
```
$ oasis network add dryrun 53852332637bacb61b91b6411ab4095168ba02a50be4c3f82448438826f dryrun.grpc.oasis.dev:443
? Description: bla
? Denomination symbol: DRY
? Denomination decimal places: 9
Error: malformed chain context: encoding/hex: odd length hex string
```

File: cli/cmd/network.go
Comment (tjanez) [inline cli/cmd/network.go:66]: There is a bit of an inconsistency. In the `oasis network list` we use `NAME` and here we ask for `Description`.
I'm slightly more in favor of using `Name` here and perhaps `network_name` for the variables to avoid confusion with all other `Name` things in this part of the code :slightly_smiling_face:

File: cli/cmd/network.go
Comment (tjanez) [inline cli/cmd/network.go:106]: This doesn't need to be resolved in this PR, but could the shell completion be configured to list all the networks' names when one presses <kbd>Tab</kbd> after typing:
```
oasis network rm
```

File: .gitignore
Comment (kostko) [reply .gitignore:27]: Done.

File: cli/go.mod
Comment (kostko) [reply cli/go.mod]: Done.

File: cli/cmd/network.go
Comment (kostko) [reply cli/cmd/network.go:106]: Yeah this should be done in a separate PR, feel free to file an issue once this is merged.

File: cli/README.md
Comment (tjanez) [inline cli/README.md:69]:
```suggestion
All configuration is stored in the `$XDG_CONFIG_HOME/oasis` directory (defaults to `$HOME/.config/oasis`).
```

File: cli/cmd/root.go
Comment (tjanez) [inline cli/cmd/root.go:39]: I would perhaps go with `cli.toml` if we ever add more configuration files under `$XDG_CONFIG_HOME/oasis`?

File: cli/cmd/root.go
Comment (tjanez) [inline cli/cmd/root.go:44]: Do you think TOML will be better for the users :slightly_smiling_face: ?
I'm a bit more in favor or YAML :slightly_smiling_face:

**Comment by Yawning** [comment]
Since I'm blocked on this, I might as well request changes too:
 * Can you move cli/client to somewhere inside the actual client SDK so that people writing tooling aren't stuck either reimplementing it or pulling it in from the CLI.
 * Likewise, can you move cli/common/demonimation to somewhere inside the actual client SDK?

**Comment by tjanez** [comment]
> Since I'm blocked on this, I might as well request changes too:
> 
>     * Can you move cli/client to somewhere inside the actual client SDK so that people writing tooling aren't stuck either reimplementing it or pulling it in from the CLI.
> 
>     * Likewise, can you move cli/common/demonimation to somewhere inside the actual client SDK?

Yeah and perhaps extract this in a separate commit and PR so it can be reviewed and merged earlier?

File: cli/cmd/network.go
Comment (kostko) [reply cli/cmd/network.go:66]: So there are two things, "name" and "description". The name is an identifier used in various places to reference networks, paratimes etc. while "description" is something human readable that is also shown when signing a transaction.

File: cli/cmd/network.go
Comment (kostko) [reply cli/cmd/network.go]: Done.

File: cli/cmd/root.go
Comment (kostko) [reply cli/cmd/root.go]: Done.

**Comment by kostko** [comment]
Moved some of the config, connection and helpers stuff to the Go Client SDK.

**Comment by kostko** [comment]
I'll merge the PR and we can improve the CLI in subsequent PRs.

File: client-sdk/go/config/default.go
Comment (pro-wh) [inline client-sdk/go/config/default.go:11]: resposting as this was lost in a file move:

        does the grpc-web proxy lets us dial it from the normal grpc client? or is there an additional thing on that server that handles normal grpc?

**Pull request merged**
